Execution Management of Controllers (#8)
ros-controls/roadmap: design_drafts/controller_execution_management.md

# Controller Execution Management

The following shall discuss the execution of loaded controllers inside the `ControllerManager`.
As of today, the loaded controllers are processed in a sequential order which doesn't allow for much flexibility or interoperability between the controllers.
We identified a few essential criteria when it comes to the execution of controllers:

* Sequential Execution of Controllers
* Parallel Execution of Controllers
* Nesting Controllers
* Variable Frequencies of Controllers

We'll go into more details below for each aspect.

## Prerequisites

### Flexible Joint State Message

Over the course of this document, we are referring to a "Flexible Joint State Message".
This data type is precisely described in [here](flexible_joint_states_msg.md).

What is important to understand for this article is that this joint state message serves as a general container structure for various, dynamically specified key values pairs.

> Every controller can hereby access values which are specific to itself without being limited to a set of pre-defined keys.

Reviewer: Do we need to explain how the controllers have access to them here? (ex. IPC, messages...?)

Author: I consider the article about the `flexible joint state message` currently as incomplete and would prefer having that sort of discussion in a new PR for that article.
But in principle, I think it should stay somehow opaque to the controller what sort of message is present and how the input/output keys are connected. I think it's sufficient for an individual controller to read and write custom fields in this message. Similarly, they should not be concerned with access control in a sense that each controller has full control over the associated keys.
Does this make sense?

For example can a `JointEffortController` access the robot's input state via keys like `joint_position` and `joint_velocity` and write its output torque under a key such as `torque_command`.

> This further allows an easy extensibility in the variety of controllers which could access data completely independent of their underlying hardware.

Reviewer: allows ease of extensibility for various controllers that can access data...

> Various components such as camera systems or third-party peripherals can therefore be combined and shared between controllers.

Reviewer 2: I meant *joint* state :)

A typical example for this are visuomotor controllers where camera images are used to generate torque values.

> Relevant features extracted from camera inputs can so be shared between multiple controllers to eventually calculate the appropriate control command.

Reviewer: can ~~so~~ shared...

### Controller and Hardware Interface

Each controller gets the previously mentioned joint state message attached during their initialization phase.

> The controller manager is responsible to initialize the joint state message appropriately and route it correctly to each controller.

Reviewer: * for initializing...
* routing it...

> It is further the task of the controller manager to fill the message with a call to `read` from the connected hardware interface and apply the results from the controllers accordingly with a call to `write`.

Reviewer: Does it do this by passing the setpoint to each individual controller or should we account for a case where some calculation needs to be performed?

Author: Not sure I understand your comment. Can you please elaborate?

What I tried to say here is that the controller manager is sole entry point for communicating with the actual underlying hardware. The robot's state can be fetch by a call to `read() -> flexible_joint_message` which creates/fills the joint state message. It is then passed to the controllers, which perform an update and the final joint state message is then being passed to the hardware to apply their values. All calculations/modifications to the map shall be done in the controllers exclusively.

Reviewer: Thanks for the clarification!
It seems my concern here is more applicable to my question on parallel controller container frequencies below and is unrelated to the responsibility of the controller manager.

Another use case could be to clamp a `DiffDriveController` velocity values to optimize for battery lifetime or top acceleration.

In order to realize such a controller chaining, the output of one's controller must be linked as the input to its consecutive controller.
This further motivates a relaxation that resources such as the `JointHandle` can only be uniquely claimed.

### Sequential Controller Container

Instead of loading each controller individually as an independent controller, we propose a parent container which holds and connects controllers which are configured in a specific order.
The container has a single input and a single output which is being scoped within the container.
That is, all controllers loaded within a container don't see any input/output states outside the container.
The input and output states can then be safely re-routed and linked between individual controllers.
After all controllers are executed, the final output state is being returned from the container.

The image below shows how this execution model could look like:
![sequential_execution](images/sequential_execution.png "Sequential Execution Container")

## Parallel Execution of Controllers

> Opposite to the sequential order, there are use cases where it makes sense to operate controllers in a parallel fashion.

Reviewer 2: ```suggestion
Complementary to sequential execution, there are use cases where it makes sense to operate controllers in a parallel fashion.
```

> `Controller_1`, `Controller_2`, …, `Controller_3` are being executed in parallel/isolated and their output state is being fused with a specific function.

Reviewer 2: ```suggestion
`Controller_1`, `Controller_2`, …, `Controller_3` are being executed in parallel, each operates on a copy of the group's input and individual output states are being fused with a specific strategy.
```

> `Controller_1`, `Controller_2`, …, `Controller_3` are being executed in parallel, each operates on a copy of the group's input and individual output states are being fused with a specific strategy.

Author: ```suggestion
`Controller_1`, `Controller_2`, …, `Controller_N` are being executed in parallel, each operates on a copy of the group's input and individual output states are being fused with a specific strategy.
```

### Separation of Tasks

Having controllers being run in parallel allows a better separation of tasks per controller.
One easy to grasp example (also if not the most sensible) is a simple `PID-Controller`.
We could break down each gain (`P`, `I`, `D`) in its own controller and let them calculate in isolation.
By being able to execute multiple controllers in parallel, the calculation of the actual output can then be to simply add the `P`, `I`, `D` outputs - according to the control law of `p * (x_d - x) + p * (x_d - x) / dt  + i * sum(x_d -x) * dt`.

> A second use case could be again a differential drive controller which computes rotational and translational velocities and a second controller calculates dynamic friction values which then again can be fused together.

Reviewer 2: ```suggestion
A second use case could be a differential drive controller which computes rotational and translational velocities and a second controller calculates dynamic friction values which then again can be fused together.
```

### Parallel Controller Container

Just like in the sequential execution, we propose having a container structure which holds a set of loaded controllers which are being executed in parallel and primarily independent from each other.

> When all controllers were executed, each individual calculated values are being joined.

Reviewer: If the parallel controllers are operating at different frequencies, do we need to be concerned about the rate at which the container is operating at?

Author: very good point. I have thought about this, but couldn't really come up with a good solution. We could consider having each controller adhering to the container's frequency, which would make it cumbersome to cluster controllers with different frequencies. 
We could alternatively giving each container it's own instance of an executor which handles the frequencies. That would lead to what I'd call "nested frequencies" where the absolute frequency specified for a controller depends on the container's base frequency. 
Yet an alternative approach could be to update each controller in its desired frequency and only run the "fuse" operation at the container's frequency.

I have to think more about all these approaches and their caveats. Any ideas are merely appreciated.

Reviewer: This is starting to sound similar to the QoS policy design with rosbag2 but with frequencies :)

> ...giving each container it's own instance of an executor which handles the frequencies.

I'm in favor of this approach but I don't know of any repercussions off the top of my head. Maybe added complexity? It does need some thought....
I'm not familiar with what you mean by fusing though.

Reviewer 2: Perhaps in the case of a parallel cluster of controllers, variable update rates can be supported by setting a fixed frequency to the fuse operation, a sort of heartbeat.

> As already mentioned previously, this parallel container can be initialized with a custom designed operation which dictates how all commonly calculated values (e.g. `torques`) are being combined.

Reviewer 2: ```suggestion
As already mentioned previously, each controller group type can be initialized with a custom strategy which dictates how all commonly calculated values (e.g. `torques`) are being combined.
```

Author: This strategy is not applicable for a sequential controller group, but exclusively for the parallel one.

The parallel container also has only one input and one output state.
However, the input state has to be multiplexed (essentially copied) for each controller and can't be shared in order to guarantee and isolated execution where no other controller within the same container can interfere.

The image below shows how this execution model could look like:
![parallel_execution](images/parallel_execution.png "Parallel Execution Container")

> ## Nested Controllers

Reviewer 2: This sounds great in theory but I wonder how much value we'd get out of it. It definitely opens another set of possibilities, somewhat similar to sequential controller groups but it may get pretty hard to trace down any issues within a nested system of systems. While I like this idea too, I think it should sit further down the priority list.

Author: We actually have a very strong use case for this. In principle every cascading controller scheme would need this feature. Therefore, at least for us, that's one of the major improvements we'd like to see in ros-control.

Author: https://www.controleng.com/articles/fundamentals-of-cascade-control/

In both cases, the sequential as well as the parallel container can be controllers in itself.
That gives the flexibility to easily nest the container within another container transparently.
Again, a very simple scenario could be a parallel `PID` controller setup which is then being configured within a sequential container to enforce joint limits.
This further allows to consider maybe very complex configured containers as black boxes, yet have the freedom to extend these fairly easy.

The image below shows how this execution model could look like:
![nested_execution](images/nested_execution.png "Nested Execution Container")

## Variable Controller Frequencies

One big constraint of the round-robin approach is the constant update frequency for all loaded controllers.

> There are plenty of use-cases however where it makes sense to have variable update frequencies, such as running a robot controller with 1 kHz whereas a camera controller only receives new images with a 20 Hz frequency.

Reviewer 2: Let's keep it general
```suggestion
There are plenty of use-cases however where it makes sense to have variable update frequencies, such as running a robot controller with 1 kHz whereas a controller operating on a slower sensor only receives new data at e.g. 20 Hz.
```

Every controller should therefore be attached with its own update frequency.

Looking at the `rclcpp executor` model, one approach could be to start a timer for each controller where its callback will call the respective `update()` function.
In order to establish a deterministic behavior of the timers, the controller manager has to make sure that really only timers are started and no other blocking calls such as subscriptions or services are part of it.
There is quite some discussion about this at the moment of writing here: https://answers.ros.org/question/327477/ros2-uses-6-times-more-cpu-than-fastrtps/

The potential options are having a separate executor designed for ros control with adheres to the interface of the `rclcpp::executor` or to neglect the concept and implement a custom execution model for ros_control.

> ## Implementation - Proof of Concept

Reviewer 2: Let's add a new section or something to this elaborating on ideas for existing/new tooling that could be used to assess the state of a system in configuration time, runtime, etc.

For configuration I can see a tool reading configuration files from a launch file perhaps that was meant to bring everything up (this way not all controller config files need to live in the same place) and construct a schematic of the system with a simulink-style gui?

Author: yes! I very much like the idea to have some sort of human readable configuration file (.yaml) or such which designs the controller setup. That would also allow to even have bindings to graphical user interfaces which generate these config files.
